### ISSUE TITLE
Handicap training

### ISSUE THREAD

**Issue by Dorus** [gcp/leela-zero#1313]
There was some talk about training a handicap NN in the other issue, however it's already quite long and handi games are offtopic there too. So want to continue that part of the discussion elsewhere. 

https://github.com/gcp/leela-zero/issues/1311#issuecomment-386282213
> herazul commented 9 hours ago
>
> I think the most interesting thing by far would be to train  a network that can handle several stones handicaps and weird komi, but i  have no clue how. Seems like Golaxy kinda managed to do it, but i have  no idea how.

https://github.com/gcp/leela-zero/issues/1311#issuecomment-386283469
>  gcp commented 9 hours ago 
>
> Yes, this is something that would work. If it is the best way to do it, I do not know. Training with (high) handicap is trickier, as you somehow need to  ensure the games are not too lopsided. So you need imbalanced playouts,  or different temperature, or something.

https://github.com/gcp/leela-zero/issues/1311#issuecomment-386287315
>  Eddh commented 9 hours ago •
>
> For a next run I agree that the most interesting things to try would be variable komi and handicap. Following @bittsitt idea, it should be possible to have 1 input plane for komi 7.5, 1 input plane for komi 6.5, 1 input plane for komi 0.5 etc ?
> 
> For handicap games, would it be possible to have more random conditions for the game, in addition to the random komi ? Placement of random stones on the board, somewhat randomized playout count, etc. Randomized training condition have already been used in neural network training https://www.youtube.com/watch?v=9EN_HoEk3KY&t=23m36s so that the neural net can better adapt to new conditions.
> 
> Also, so that the neural net can take into account the end score rather than just win/loss, is it possible to score a games as 1.15 if it is won by resign, but also 1.15 if it is won with a score that is more than 15. That way the system would want to resign if it is losing by more than 15 points difference, but carry on otherwise, which would be similar to human play.
> 
> I am aware that all the pretty ideas will need work and expertise to be implemented at some point though.

IMHO training with handicap should be easy if you are also willing to train with different komi values. Just start off training with 9 handicap stones but -150 komi to compensate, or 5 handicap stones and -75 komi etc. It might take some fiddling to find the right komi for each handicap value, but eventually you would want to train the value net for any komi. I'm not even convinced komi needs to be a network input, it could indeed be added at the policy and value block only. When playing against handicap stone, the bot could start at -15 komi for each handi stone, and slowly switch up to the required komi each time the win rate for the next komi got into acceptable range. MCTS is said to work best in the 40-60% range, so it could start at -75 komi and switch to -60 komi once that goes into the 40% range (or we might require smaller steps, but the idea stays the same).

My basic idea is that against handicap, the perfect game should result in something like -150 points with 9 handi games. The NN only needs to play this perfect game, only when the opponent makes a mistake it should pick up points. you can still tune how aggressively you push towards the played komi of 0.5, the more aggressively you push towards this, the more likely the bot will overplay.

**Comment by gcp** [comment]
But then your bot needs some way to know the exact komi used? And you need to train and vary that *too*, in addition to the handicap? What is the advantage of this proposal over the others that were made? 

As for other disadvantages:

>I'm not even convinced komi needs to be a network input, it could indeed be added at the policy and value block only.

But being ahead of behind should have a massive influence on the policy selection, and you can't reasonably expect the network to do something meaningful with it if it skips 99% of its knowledge.

If we change komi to 5.5, it is not unreasonable to expect the fuseki and joseki that are discovered to change. So IMHO adding komi *and* handicap would really stress what one network has to deal with. (I pointed out that earlier experiments with multi-labeling split the networks, so this stress on training the value network does not affect the policy network. But of course, that approach has other issues)

>MCTS is said to work best in the 40-60% range

I am not sure this applies to neural networks at all. The 40-60% range is very good for binominal functions as this is the place of maximum entropy per simulation, but our neural network eval is not really that as it gives continuous results.

**Comment by odeint** [comment]
Could the easiest way to train a player who maximises territory from a given situation, like most humans do? That way, a handicap game can just be played normally, and ends with a score difference that is the value of the handi stones at the given level of play. What were the problems with this approach? Losing when ahead due to greed maybe, but once the network is super strong in self play it will know that overplays do not generate extra territory, but lose it. It's only a problem against a stronger opponent.

**Comment by herazul** [comment]
I am pretty noob at machine learning, but i was wondering something : 
human can be good at handicap games because we see more things that just winrate of a position on the board. We can take a lossing board position versus a lower level player and think something like "here i have secured points, here and here there is weaknesses in shape that i can try to exploit to get back points, ... so i'm gonna play this and this risky move and try to exploit every weekness i can with my superior reading."
As i understand it, on the same loosing position on the board, the network is gonna evaluate a good move and be like "nope, still a loosing position" even if it can be a good move to catch up with god-reading abilities.
So my point is : is it really on the net and training that Golaxy did things differently, or maybe in the engine there could things like going for more aggressive and combat play if it's behind ? I don't really know how it would be possible, but maybe there is a way to exploit a god-tier network ability to read positions and make it play aggressive moves and combat moves.

**Comment by marcocalignano** [comment]
I my opinion, since on a Handicap game you play different than usually, you know your opponent is less strong, so you make move that are not the best to confuse your opponent and the get your advantage, one solution would be to train a different network with handicap games. Then we could use the "normal" one in an even game, and the "handicap" one for handicap game. A bit like a human does.

**Comment by herazul** [comment]
Would it be possible to code a search function that try to take into account points AND winrate and not just winrate and try to maximise points and not just winrate with current network ? or is that theoretically impossible ?

**Comment by gcp** [comment]
I did this for Leela but I've never seen much evidence that it's really necessary for LZ. It's more of a "cosmetic" issue. I don't think it's interesting. Points are not the goal of the game. (And this really doesn't change when taking handicaps).

It's useful for analysis tools, but I already explained how you can train a network to give you a count.

**Comment by gcp** [comment]
>Could the easiest way be to train a player who maximises the expected territory from a given situation, like most humans do?

This makes the bot significantly weaker. (See: tons of research already published on this)

**Comment by gcp** [comment]
>I my opinion, since on a Handicap game you play different than usually, you know your opponent is less strong, so you make move that are not the best to confuse your opponent and the get your advantage, one solution would be to train a different network with handicap games. Then we could use the "normal" one in an even game, and the "handicap" one for handicap game. A bit like a human does.

I don't think it is necessary. The network can see that some stones are missing. Regardless if it's due to handicap or making a huge mistake in the opening, optimal play at that point relies on the opponent making a mistake. So the strategy is similar.

**Comment by gcp** [comment]
>is it really on the net and training that Golaxy did things differently, or maybe in the engine there could things like going for more aggressive and combat play if it's behind 

I don't know what Golaxy does but I would recommend trying to propose a solution that helps the network understand the situation, or it will never be useful in a LZ/AZ style setup.

**Comment by jkiliani** [comment]
> I don't think it is necessary. The network can see that some stones are missing. Regardless if it's due to handicap or making a huge mistake in the opening, optimal play at that point relies on the opponent making a mistake. So the strategy is similar.

I'm not sure this works with the current setup. When I played Leela Zero with handicap 4, it always does every 3-3 invasion it can immediately, which is a terrible idea for white in a handicap game. Leela 0.11 on the other hand starts fights immediately by playing on the edges and towards the centre, and it wins those fights very regularly.

I think the problem is that the current networks assume the opponent will never make mistakes, which makes negating a handicap hopeless. Likely, either a temperature-based solution or a MCTS rollout-based one would work here...

**Comment by herazul** [comment]
Okay thanks a lot for your answer, that makes a lot of sense.
So maybe the fact that the net train with it self on optimal even play is just what make it weaker on handicap games comparing to a net that train on weaker games and human games and see a lot of "weird" board positions.
If i understand it correctly then, we could trade pure even match strength for better handicap strength if we took the last net and do a training session on a set of human games. Maybe even a pass on a selection of interesting handicap games.
But all in all, that make what Golaxy did even more mysterious.

**Comment by fell111** [comment]
Maybe Golaxy trained another network for komi difference? That's only my guess. 
Is that possible to train such a net and using together with the winrate net, just like value network and policy network in the first paper of AG?

**Comment by herazul** [comment]
@marcocalignano Yep that would not be in official training obviously, but that would indeed be a way to trade some pure strength for some handicap strength. The way we could do it is to generate handicap games along the way, and when we have enough we take the last net and train it on this set of handicap games.
But to have a meaningfull set of handicap games, we would need to be quite a few to generate these games, i don't know how we could set that up.

**Comment by marcocalignano** [comment]
With autogtp and the server we can set it up if we want to, just create a new command for handicap games and a new url where the upload them. But the question is why not insert them in the mainstream training. Do you think would disrupt the "normal" network?

**Comment by herazul** [comment]
Yeah it would probably weaken the training and it would even pass gating anymore.

**Comment by gcp** [comment]
>I think the problem is that the current networks assume the opponent will never make mistakes, which makes negating a handicap hopeless.

Right, but that is unrelated to my argument. We can make the training games make more mistakes, especially in handicap games.

**Comment by gcp** [comment]
>Maybe we should do train the network with self play where a weaker network plays against a better network with handicap. 

Seems more easier (and faster) to reduce playouts for the side receiving the handicap. Doesn't require fiddling multiple networks.

AutoGTP would need support for that though, currently self-play uses 1 engine only and you can't vary options.

**Comment by gcp** [comment]
>trade some pure strength for some handicap strength
>Yeah it would probably weaken the training and it would even pass gating anymore.

I am not convinced this would even be an issue. As explained above, the network may very well learn to play better when behind in an even game.

**Comment by jkiliani** [comment]
> Seems more easier (and faster) to reduce playouts for the side receiving the handicap. Doesn't require fiddling multiple networks.

While the idea is interesting, training data in such self-play games would be limited to only the side giving the handicap I presume?

**Comment by marcocalignano** [comment]
> AutoGTP would need support for that though, currently self-play uses 1 engine only and you can't vary options.

AUTOGTP needs to support it anyway. In the "same network" solution needs also to start two leelaz processes with different playouts, set the handicap stones. In the two network solution also he needs to start two leelaz processes with two diffenrant networks and set up the handicap stones. The only difference is that with the playout you do not to choose the second network (easier) but you might loose in diversity.

**Comment by gcp** [comment]
>While the idea is interesting, training data in such self-play games would be limited to only the side giving the handicap I presume?

Bah, that's a good point. It would mess up the value head? Suddenly, the expected score for winning/losing a stone early on is 50%. I think that problem exists regardless of which side you train.

**Comment by gcp** [comment]
I think it also undermines my point about using the same network. In a handicap game, if you are behind a stone you want the score to stay reasonable. In a normal game, if you see a way to win a stone you want a high score.

So at the very least, trying to equalize the winrates in the handicap games by handicapping one side in terms of strength could badly backfire because it introduces a consistent bias into the value network.

Just increasing the temperature may be more beneficial in terms of getting good behavior on the same network.

**Comment by odeint** [comment]
Golois6 on KGS seems to be doing quite well against handicap, regularly winning H6 komi 0.5 games against 1 dans there (being high 7d itself). In the description it says:

> I use a policy network trained on Gogod games and a value network trained from self play.
> 
> If I don't mark dead stones as dead in the end of the game, please undo and capture the stones.
> 
> Players using a wrong handicap will be added to the censor list.
> 
> Thanks for playing with me :)

So this extra policy network seems to help it a lot in achieving good performance against handicap.

**Comment by gcp** [comment]
That's basically the Lee Sedol architecture which Leela 0.11 also uses. As noted in the other thread, it does work better with handicap.

**Comment by Marcin1960** [comment]
What about equalizing the winrates by alternating advantage to the sides during the training?

**Comment by odeint** [comment]
> But it is not competitive strength wise.

petgo3 wins about 88% against Golois6 in even games ( http://kgs.gosquares.net/index.rhtml.en?id=petgo3&id2=Golois6&r=1 ).

Stronger, but not totally outclassed. Is it possible to have just one network (dual head), but training an extra policy head on something like the GoGoD dataset? Could that get close to the performance of a complete policy network trained for just that purpose?

**Comment by gcp** [comment]
Seems problematic in the AZ/LZ architecture. At some point the GoGoD games are like feeding training from very weak networks into it. Nobody tested this of course so I don't see how to even begin predicting the performance.

**Comment by herazul** [comment]
That quite an interesting idea : a policy network that was trained on watever games to become an expert at finding weird and efficient move for handicap games, that ask the value net of the current best net to validate move.

**Comment by betterworld** [comment]
> Seems more easier (and faster) to reduce playouts for the side receiving the handicap. Doesn't require fiddling multiple networks.

Could the expected strength difference be used as an additional input for the neural network?  I think this avoids problems in the value head. Otherwise it might develop some idea of "whenever my position looks bad in the opening, my opponent will have weak reading".

**Comment by gcp** [comment]
>Could the expected strength difference be used as an additional input for the neural network?

Yes but then you could as well add an explicit plane with the handicap (stones) on it, and play towards 50% if set.

**Comment by Hersmunch** [comment]
I like the idea of increasing the temperature just for the opponents moves. I.e. what if they played an unexpected move? But then being able to focus its own moves without penalty. 
Not sure how to relate the temperature used to the number of handicap stones though. Maybe link it to the expected win rate?
Also wondering if it might be useful if playing a known stronger opponent.

**Comment by Dorus** [comment]
I was going to reply here earlier but didn't have the time. Going to add few things i've been thinking about:

My first theorem is that handicap games do no require overplay to be won. In fact, overplay, as the name suggest, makes you weaker.

If an opponent can beat your 9d with 2 handi stones, that probably means your opponent is 7d or stronger, it does not mean you should have weaker moves hoping your opponent would fail to deal with them.

The real way to win handicap games is by playing a perfect game where you give zero points away each move, and wait for your opponent to make mistakes where it does give way points. Ofc in practice both  players will make mistakes, but the stronger player generally makes fewer and thus ends up winning despite the handicap stones

My second theorem is that in order to deal with handi games, you do not so much need a weaker opponent to train against (with temperature, more noise, less play outs etc), instead it would be enough to assume a large komi that compensate the handicap stones. That way the network will be able to learn the perfect move in handicap game.

The reason you need to do something is this is that if you play with normal komi, every move will end up losing, so after the network finds the strongest possible move, analyzes it to dead, it will reject it as it was losng. However, if you let it play against a weaker opponent, it might instead settle for a weaker move that somehow exploit the given opponent without actually gaining an advantage vs other player.

Lastly, my suggestion to connect the komi only to the value and policy head was two fold. First given the above reasoning that the network always need to find the perfect move anyway, and possible the safest move too, it would be enough for the policy head to know the current komi, calculate if it needs more points, and pick between the safe move and teh best move. The value net just needs to learn to count anyway, so for the value net it would be even easier to deal with komi added only at the head. Now my second reason for this suggestion was that this would make it easier to convert the current network. I might be wrong about that, and if so, my whole idea to add komi value at the head instead of the input is bogus of course.

I also have some more ideas on how to train for handicap games, but first i'm going to study a bit more what has already been suggested :-)

**Comment by gcp** [comment]
>The real way to win handicap games is by playing a perfect game where you give zero points away each move, and wait for your opponent to make mistakes where it does give way points.

I have trouble wrapping my head around what this means in a fighting game.

I think your argument also assumes there's no such thing as complicating the situation and making it more likely for a weaker player to mistake. I'm not sure it's realistic to ignore that.

>Where do we set the temperature to anything other than max i.e 1?

It's an exponentiation operator on probabilities. So that just works.

**Comment by PhilipFRipper** [comment]
What about scaling win factor up at start of a handicap game by X and having it aneal to default over, say, 100 moves.  So that it is fooled into thinking it is competitive early but requires the position for itself to improve as the game continues to continue to avoid resigning.  I don't know if that's too hacky.

**Comment by l1t1** [comment]
I wonder the handicap x isn't equal to force leelaz pass first x move and then run as normal?

**Comment by wonderingabout** [comment]
Maybe FineArt B/C on fox can be a source of inspirations ?

It used to give up to 4 stones to 9d players with > 90% winrate

**Comment by Dorus** [comment]
@Ishinoshita But going so would make the bot weaker unnecessarily. Of course you could make a "maximize score" setting where it keeps raising the internal komi forever, but remember each time you do, it will resort to making the game more complicated and try more daring moves. A perfect net wont make any mistakes in this, but a weaker net greatly benefit from the MCTS property that once it has enough points, it will play the safest move, not the largest move. If you keep raising internal komi, it will continue to play the largest and not the safest move.

**Comment by Hersmunch** [comment]
How about a policy head (and value head) for each selected Komi being trained? Then each can select a safe but large enough move for its Komi.

**Comment by gcp** [comment]
> for vanilla MCTS

As already pointed out, there is a large difference here between A(G)Z and vanilla MCTS. Vanilla MCTS is a binominal distribution that can gather no information if p is close to 0 and 1. But the A(G)Z design gives a continuous score, where this isn't a problem at all.

**Comment by dorzechowski** [comment]
Would it be possible to schedule a match against some old network on 1 and perhaps 2 stones (i.e. 2 or 3H with komi for White)? It would be good to know where we stand in view of upcoming handicap matches. 

I think (looking at only few such games myself, so no confidence in this claim) that we need around 1000 self Elo advantage for 1 stone handicap. On 2 and 3 stones is hard to even guess because Leela plays all possible 3-3 points right away and the higher handicap the worse idea it is.

**Comment by roy7** [comment]
Current setup doesn't support handicap at all, but one could maybe customize the gomill software to start a match by giving black an extra play first if the idea is just to test the win rate during a handicap situation.

**Comment by alreadydone** [comment]
Comments welcome for my PR https://github.com/gcp/leela-zero/pull/1435
(not retraining the NN but modify the tree search for handicap play)

**Comment by Marcin1960** [comment]
@alreadydone "Comments welcome for my PR #1435"

Running your change on KGS as LeelaZeroT

**Comment by odeint** [comment]
> Running your change on KGS as LeelaZeroT

Nice, how many handi do you allow max?

**Comment by Marcin1960** [comment]
9 stones

**Comment by alreadydone** [comment]
Newly promoted network has [higher winrate](https://github.com/gcp/leela-zero/issues/1405#issuecomment-389029467) for handicapped games; looks like fixed Dirichlet noise and t=1 all game had some effect.

**Comment by roy7** [comment]
So it sounds like, from the point of view of win rate not being overly compressed to 0% to mess up MCTS search, LZ can play 3-stone better than the other networks and it's getting better as time passes?

LZ-ELF at .18% win rate would just melt down with 3-stone.

**Comment by Hersmunch** [comment]
I stumbled across this [AI Safety Gridworlds paper](https://arxiv.org/pdf/1711.09883.pdf) from DeepMind that I think briefly discusses two topics to what we're interested in, in these sections:
wrt handicap
> 2.2.2 Distributional Shift
> Deep reinforcement learning algorithms are insensitive to risk and usually do no cope well with
> distributional shifts (Mnih et al., 2015, 2016). A first and most direct approach to remedy this situation
> consists in adapting methods from the feedback and robust control literature (Whittle, 1996;
> Zhou and Doyle, 1997) to the reinforcement learning case (see e.g. Yun et al., 2014). Another
> promising avenue lies in the use of entropy-regularized control laws which are know to be risksensitive
> (van den Broek et al., 2012; Grau-Moya et al., 2016). Finally, agents based on deep architectures
> could benefit from the incorporation of better uncertainty estimates in neural networks (Gal,
> 2016; Fortunato et al., 2017).

wrt weaker (or stronger) opponents:
> 2.2.3 Robustness to Adversaries
> The detection and exploitation of the environmental intentions has only recently drawn the attention
> of the machine learning community. For instance, in the context of multi-armed bandits, there has
> been an effort in developing unified algorithms that can perform well in both stochastic and adversarial
> bandits (Bubeck and Slivkins, 2012; Seldin and Slivkins, 2014; Auer and Chao-Kai, 2016);
> and algorithms that can cope with a continuum between cooperative and adversarial bandits (Ortega
> et al., 2015). These methods have currently no counterparts in the general reinforcement learning
> case. Another line of research that stands out is the literature on adversarial examples. Recent research
> has shown that several machine learning methods exhibit a remarkable fragility to inputs with
> adversarial perturbations (Szegedy et al., 2013; Goodfellow et al., 2014); these adversarial attacks
> also affect neural network policies in reinforcement learning (Huang et al., 2017a).

When I've got a spare moment I'll have a dig through those references..

**Comment by qrss** [comment]
Since it has been suggested that the fact that handicap games have a different ratio of stone than usual might confuse the network, perhaps we could try to make the network more invariant with respect to the ratio of stones.

For example, a small number of training games could start with a black stone randomly placed somewhere on the board.

**Comment by tapsika** [comment]
Today's collapse against Haylee seems to be worth a closer look, like at heatmaps for the weirdest moves, and at value output before/after them. I assume the policy still prefers meaningful moves, and it's the value net that breaks the search?

**Comment by yeechutngeow** [comment]
I am wondering whether some form equalizing of the winrate by some form of up-scaling would prevent LZ from going crazy like it did in the Haylee match.  Maybe a function that is proportional to the handicap stones and inversely proportional to number of moves into the game.

**Comment by Hersmunch** [comment]
So I've had a quick skim of the first few papers mentioned [above](https://github.com/gcp/leela-zero/issues/1313#issuecomment-392322837). A few take away thoughts include using risk estimation/measure during search to adapt the policy based upon the observed opponent's behaviour in terms of differences between what move they make compared to the moves we predicted they would make. The opponents behavior might be better modelled with an uncertainty parameter. This could allow for risk-averse pessimistic behaviour when more certainly ahead and risk-seeking optimistic behaviour when behind.
I have to say that most of the papers are outside of what I can understand so please see what you can make of them too.

**Comment by alreadydone** [comment]
Latest network # 148's winrates against fixed handicap stones are roughly 54%, 28%, 12%, 2%, 0.5%, 0.1%, 0.05%, 0.05%, 0.05% (9-stone) after several hundreds of playouts. With 9 free handicap stones it's about 0.2% (>0.05%). Though the winrates increased, new networks don't distinguish as clearly among 7,8,9 handicap stones as earlier networks, so it's not clear whether handicap play will be better. For comparison, ELF distinguishes poorly among 3-9 (or more) handicap stones. Notice that [ELF did use the AZ approach in the end](https://github.com/gcp/leela-zero/issues/1311#issuecomment-396089778).

**Comment by alreadydone** [comment]
"Dynamical komi" to keep winrate within a certain range now implemented in https://github.com/gcp/leela-zero/issues/1599.
